### DISCUSSION TITLE
Visual center vs geometric center

### DISCUSSION THREAD

**Discussion by ojwebb** [ImageMagick/ImageMagick#4835] [Help]
Does anyone know how to find the visual center of a logo based on the shape and symmetry rather than the centroid or gravity. I'm using im7, thanks guy

**Comment by fmw42** [comment]
"Visual Center" is not mathematical.  Tell us how you define that and show and example.

**Reply by ojwebb** [reply]
![image](https://user-images.githubusercontent.com/62193597/153725748-67c115be-e39c-44b7-871a-37da9abca283.jpeg)

Here's an example. This is very bottom heavy so visually looks off center if you use the geometric center. This is a similar method used in browser.

**Reply by ojwebb** [reply]
![image](https://user-images.githubusercontent.com/62193597/153725861-bfb303a7-27c2-43b3-9935-d979f09ebfc8.jpeg)

This would be a more extreme example

**Reply by fmw42** [reply]
In your first example, to me the visual center is the centroid.  In the second example, the visual center you show is just the centroid of the non-white parts.  Have you considered just using that?

**Comment by snibgo** [comment]
@ojwebb: You are using a program to find the "visual center". What program? Does its documentation explain what they mean by "visual center"?

**Reply by ojwebb** [reply]
Here is a brief description of his method.

It calculates a visual weight value for every pixel on the image: the square root of the difference between that pixel and the background color. And then the algorithm tries to balance that weight in all directions.


https://javier.xyz/visual-center/

This is the guy that inspired the idea of finding a im equivalent solution.

**Reply by fmw42** [reply]
Where did you find that description on his web site? I do not see it.

**Reply by fmw42** [reply]
Can you post the individual original image of the triangles so we can test with it?

**Reply by ojwebb** [reply]
import base64ImageUtils from 'base64-image-utils';

const _ = require('lodash');
const { base64ImageToRGBMatrix } = base64ImageUtils;

const COLOR_DIFF_WEIGHT_EXPO = 0.333;

const ROUNDS = 250;
const SIZE = 420;

function visualCenter(base64, callback, opts = {}) {
  base64ImageToRGBMatrix(
    base64,
    (err, rgbMatrix) => {
      const height = rgbMatrix.length;
      const width = rgbMatrix[0].length;
      const bgColor = normalizeColor(rgbMatrix[0][0]);

      const { visualLeft, visualTop } = calculateVisualCenter(rgbMatrix);

      callback(null, {
        visualTop: visualTop,
        visualLeft: visualLeft,

        bgColor: bgColor,
        width: width,
        height: height
      });
    },
    {
      size: SIZE
    }
  );
}

function calculateVisualCenter(rgbMatrix) {
  var visualLeft = 0.5;
  var visualTop = 0.5;

  var { visualLeft } = recursiveGetCoord(
    rgbMatrix,
    visualLeft,
    visualTop,
    'X',
    1 / ROUNDS
  );
  var { visualLeft } = recursiveGetCoord(
    rgbMatrix,
    visualLeft,
    visualTop,
    'X',
    -1 / ROUNDS
  );
  var { visualTop } = recursiveGetCoord(
    rgbMatrix,
    visualLeft,
    visualTop,
    'Y',
    1 / ROUNDS
  );
  var { visualTop } = recursiveGetCoord(
    rgbMatrix,
    visualLeft,
    visualTop,
    'Y',
    -1 / ROUNDS
  );

  return { visualLeft, visualTop };
}

function recursiveGetCoord(
  rgbMatrix,
  visualLeft,
  visualTop,
  currentAxis,
  stepSize
) {
  const bgColor = normalizeColor(rgbMatrix[0][0]);
  const height = rgbMatrix.length;
  const width = rgbMatrix[0].length;

  var visualLeftToApply = visualLeft;
  var visualTopToApply = visualTop;

  const ops = {
    bgColor,
    height: rgbMatrix.length,
    width: rgbMatrix[0].length,
    maxDiff:
      Math.max(bgColor.r, 255 - bgColor.r) +
      Math.max(bgColor.g, 255 - bgColor.g) +
      Math.max(bgColor.b, 255 - bgColor.b),
    maxDistance: getDistance([0, 0], [width, height])
  };

  var newVisualLeft = visualLeft;
  var newVisualTop = visualTop;

  if (currentAxis === 'X') {
    newVisualLeft += stepSize;
  } else {
    newVisualTop += stepSize;
  }

  var oldCenterIntensity = getCenterIntensity(
    rgbMatrix,
    visualLeft,
    visualTop,
    ops
  );
  var newCenterIntensity = getCenterIntensity(
    rgbMatrix,
    newVisualLeft,
    newVisualTop,
    ops
  );

  while (newCenterIntensity > oldCenterIntensity) {
    visualLeftToApply = newVisualLeft;
    visualTopToApply = newVisualTop;

    if (currentAxis === 'X') {
      newVisualLeft += stepSize;
    } else {
      newVisualTop += stepSize;
    }
    oldCenterIntensity = newCenterIntensity;
    newCenterIntensity = getCenterIntensity(
      rgbMatrix,
      newVisualLeft,
      newVisualTop,
      ops
    );
  }

  return {
    visualLeft: visualLeftToApply,
    visualTop: visualTopToApply
  };
}

function getCenterIntensity(rgbMatrix, visualLeft, visualTop, ops) {
  const { bgColor, height, width, maxDiff, maxDistance } = ops;

  const centerCol = visualTop * height;
  const centerRow = visualLeft * width;
  const centerPoint = [centerCol, centerRow];

  return _.reduce(
    rgbMatrix,
    (resRow, row, rowIdx) => {
      return (
        resRow +
        _.reduce(
          row,
          (resCol, col, colIdx) => {
            const cellColorDiff = rgbDiff(bgColor, col, maxDiff);

            if (!cellColorDiff) return resCol;

            const cellDistance = getDistance(centerPoint, [rowIdx, colIdx]);
            const cellColorWeight =
              cellColorDiff *
              Math.pow(1 - cellDistance / maxDistance, 0.5) *
              1000;

            return resCol + cellColorWeight;
          },
          0
        )
      );
    },
    0
  );
}

function getDistance(pointA, pointB) {
  return Math.pow(
    Math.pow(pointA[0] - pointB[0], 2) + Math.pow(pointA[1] - pointB[1], 2),
    0.5
  );
}

function normalizeColor(color) {
  return {
    r: Math.floor(color.r * (color.a / 255) + 255 * (1 - color.a / 255)),
    g: Math.floor(color.g * (color.a / 255) + 255 * (1 - color.a / 255)),
    b: Math.floor(color.b * (color.a / 255) + 255 * (1 - color.a / 255)),
    a: 255
  };
}

function rgbDiff(baseColor, testColor, maxDiff) {
  if (testColor.a === 0) return 0;

  const diff =
    Math.abs(baseColor.r - testColor.r) +
    Math.abs(baseColor.g - testColor.g) +
    Math.abs(baseColor.b - testColor.b);
  const result =
    Math.pow(diff / maxDiff, COLOR_DIFF_WEIGHT_EXPO) *
    (testColor.a / 255) *
    1000;

  return result;
}

// module.exports = visualCenter;
export default visualCenter;

**Reply by ojwebb** [reply]
![image](https://user-images.githubusercontent.com/62193597/153733711-93d07de1-6340-427c-b70a-d2f22a00201d.png)

![image](https://user-images.githubusercontent.com/62193597/153733713-8c913ceb-d72a-4cef-b1ec-4355c6677287.png)

**Comment by snibgo** [comment]
The website says:
```
The center is at 50%, 66.66%
```
This is the centroid of a triangle, the intersection of lines from each corner to the centre of the opposite side.

The more general definition of a centroid of any shape: it is at coordinate (xbar,ybar) where xbar is the average of all the x-coordinates in the object, and ybar is the average of all the y-coordinates in the object.

That definitions assumes an equal weight for all pixels in the object. We might instead weight pixels by colour, or difference from a reference colour, or whatever. In this example, the triangle is identical in 120-degree rotations, so the result will be the same.

I wrote a process module for weighted centroids. Windows BAT script.

tri3.png is:
![tri3](https://user-images.githubusercontent.com/33812368/153733696-b974f726-6601-4f01-84a0-97f87a233c36.png)
```
magick ^
  tri3.png ^
  -background White -layers flatten ^
  ( +clone -fill White -colorize 100 ) ^
  -compose Difference -composite ^
  -evaluate Pow 0.5 ^
  -process centroid ^
  NULL:
```
The text output is:
```
centroid: 150.50002,171.29938
```
171.3 is about 2/3 of the image height.

**Comment by fmw42** [comment]
You can get the centroid of the non-white or non-transparent pixels from the Image moments from `magick identify -verbose -moments`


![Unknown](https://user-images.githubusercontent.com/13458218/153736334-cf37fcb8-0b1e-4b78-8482-0076276ab1d2.png)

`magick unknown.png -alpha extract -colorspace gray miff:- | magick identify -verbose -moments -  | grep "Centroid:"`

**Centroid: 150.501,171.469**

**Reply by ojwebb** [reply]
Wow!! That worked guys! I was always staring at the foreground. It distracted me from incorporating the background as well! Amazing logic guys!
